### PR TITLE
Simplify sorting of site package entries

### PR DESCRIPTION
**Pull request by zanieb**
per https://github.com/astral-sh/uv/pull/2780#discussion_r1548340373

### CODE REVIEW

File: crates/uv-installer/src/site_packages.rs [hunk 1]
```diff
@@ -47,11 +47,16 @@ impl<'a> SitePackages<'a> {
...
-                    let mut entries = site_packages.collect::<Result<Vec<_>, std::io::Error>>()?;
-                    // TODO(zanieb): Consider filtering to just directories to reduce the size of the sort
-                    // Sort for determinism, `read_dir` is different per-platform
-                    entries.sort_by_key(fs_err::DirEntry::path);
-                    entries
+                    // Use a `BTreeSet` to sort for determinism
+                    let directories: BTreeSet<_> = site_packages
+                        .filter_map(|read_dir| match read_dir {
+                            Ok(entry) => {
+                                entry.file_type().ok()?.is_dir().then_some(Ok(entry.path()))
```
Comment (charliermarsh) [inline crates/uv-installer/src/site_packages.rs:54]: I think this is losing `file_type` errors because they're being cast to `None`.

Comment (zanieb) [reply crates/uv-installer/src/site_packages.rs:54]: Ah yeah that is what I was having trouble avoiding.

cc @snowsignal

Comment (charliermarsh) [reply crates/uv-installer/src/site_packages.rs:54]:
```rust
let directories: BTreeSet<_> = site_packages
    .filter_map(|read_dir| match read_dir {
        Ok(entry) => {
            let file_type = match entry.file_type() {
                Ok(file_type) => file_type,
                Err(err) => return Some(Err(err)),
            };
            file_type.is_dir().then_some(Ok(entry.path()))
        }
        Err(err) => Some(Err(err)),
    })
    .collect::<Result<_, std::io::Error>>()?;
```

Comment (snowsignal) [reply crates/uv-installer/src/site_packages.rs:54]: Haha you beat me to it 😆

Comment (zanieb) [reply crates/uv-installer/src/site_packages.rs:54]: Okay yuck though is this actually easier to read than before this pull request?

Comment (snowsignal) [reply crates/uv-installer/src/site_packages.rs:54]: I tried to clean it up a bit:

```rust
let directories: BTreeSet<_> = site_packages
    .filter_map(|read_dir| match read_dir {
        Ok(entry) => match entry.file_type() {
                Ok(file_type) => file_type.is_dir().then_some(Ok(entry.path())),
                Err(err) => Some(Err(err)),
            }
        Err(err) => Some(Err(err)),
    })
    .collect::<Result<_>>()?;
```

**Review by charliermarsh — APPROVED** [review]

**Review by snowsignal — APPROVED** [review]

**Pull request merged**